server/src/main/java/io/druid/metadata/SQLMetadataConnector.java

```diff
@@ -78,7 +89,72 @@ protected String getPayloadType() {
 
   public abstract boolean tableExists(Handle handle, final String tableName);
 
-  protected boolean isTransientException(Throwable e) {
+  public <T> T retryWithHandle(final HandleCallback<T> callback)
+  {
+    final Callable<T> call = new Callable<T>()
+    {
+      @Override
+      public T call() throws Exception
+      {
+        return getDBI().withHandle(callback);
+      }
+    };
+    final Predicate<Throwable> shouldRetry = new Predicate<Throwable>()
+    {
+      @Override
+      public boolean apply(Throwable e)
+      {
+        return isTransientException(e);
+      }
+    };
+    final int maxTries = 10;
+    try {
+      return RetryUtils.retry(call, shouldRetry, maxTries);
+    }
+    catch (Exception e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  public <T> T retryTransaction(final TransactionCallback<T> callback)
+  {
+    final Callable<T> call = new Callable<T>()
+    {
+      @Override
+      public T call() throws Exception
+      {
+        return getDBI().inTransaction(callback);
```

Comment:
yes, it rollsback if any exceptions are thrown (and then re-throws the exceptions) and commits if no exceptions are thrown.